gateway/ln-gateway/src/lightning/ldk.rs

```diff
@@ -282,7 +351,50 @@ impl ILnRpcClient for GatewayLdkClient {
         max_delay: u64,
         max_fee: Amount,
     ) -> Result<PayInvoiceResponse, LightningRpcError> {
-        let payment_id = match self.node.bolt11_payment().send(
+        // Subscribe to the `PaymentSuccessful`/`PaymentFailed` events for this invoice.
+        let (sender, receiver) = tokio::sync::oneshot::channel();
+        self.payment_subscriptions
+            .lock()
+            .await
+            .insert(*invoice.payment_hash(), sender);
+
+        // Check if the invoice payment has already been started. This could happen if
+        // the gateway was stopped after the payment was initiated but before it was
+        // completed. It's important that this is done after the subscription is added
+        // to the map, otherwise we might miss the event.
+        let payment_details_or = self
+            .node
+            .list_payments_with_filter(|payment| {
+                if let PaymentKind::Bolt11 { hash, .. } = payment.kind {
+                    hash.0 == invoice.payment_hash().to_byte_array()
+                } else {
+                    false
+                }
+            })
+            .into_iter()
+            .next();
+
+        let response_or = if let Some(payment_details) = payment_details_or {
+            match payment_details.status {
+                PaymentStatus::Pending => None,
+                PaymentStatus::Succeeded => {
+                    if let PaymentKind::Bolt11 {
+                        preimage: Some(preimage),
+                        ..
+                    } = payment_details.kind
+                    {
+                        Some(Ok(PayInvoiceResponse {
+                            preimage: Preimage(preimage.0),
+                        }))
+                    } else {
+                        None
+                    }
+                }
+                PaymentStatus::Failed => Some(Err(LightningRpcError::FailedPayment {
+                    failure_reason: "LDK payment failed".to_string(),
+                })),
+            }
+        } else if let Err(err) = self.node.bolt11_payment().send(
```

Comment:
I'm finding this section hard to read. Can we make this a `match` statement? I also think the nested match can perhaps be made into a separate function. It took me a few times reading this to understand what was going on.

Reply:
Cleaned up and documented this section, and moved it into a new function called `pay_idempotent()`